File: ontrack-service/src/main/java/net/nemerosa/ontrack/service/events/EventFactoryImpl.java
package net.nemerosa.ontrack.service.events;

import net.nemerosa.ontrack.model.events.Event;
import net.nemerosa.ontrack.model.events.EventFactory;
import net.nemerosa.ontrack.model.events.EventType;
import net.nemerosa.ontrack.model.events.EventTypeNotFoundException;
import net.nemerosa.ontrack.model.structure.*;
import net.nemerosa.ontrack.model.support.Configuration;
import net.nemerosa.ontrack.model.support.NameValue;
import org.springframework.stereotype.Service;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import static java.lang.String.format;

@Service
public class EventFactoryImpl implements EventFactory {

    private final Map<String, EventType> types;

    public EventFactoryImpl() {
        types = new ConcurrentHashMap<>();
        register(NEW_PROJECT);
        register(UPDATE_PROJECT);
        register(DISABLE_PROJECT);
        register(ENABLE_PROJECT);
        register(DELETE_PROJECT);

        register(NEW_BRANCH);
        register(UPDATE_BRANCH);
        register(DISABLE_BRANCH);
        register(ENABLE_BRANCH);
        register(DELETE_BRANCH);

        register(NEW_BUILD);
        register(UPDATE_BUILD);
        register(DELETE_BUILD);

        register(NEW_PROMOTION_LEVEL);
        register(IMAGE_PROMOTION_LEVEL);
        register(UPDATE_PROMOTION_LEVEL);
        register(DELETE_PROMOTION_LEVEL);
        register(REORDER_PROMOTION_LEVEL);

        register(NEW_VALIDATION_STAMP);
        register(IMAGE_VALIDATION_STAMP);
        register(UPDATE_VALIDATION_STAMP);
        register(DELETE_VALIDATION_STAMP);
        register(REORDER_VALIDATION_STAMP);

        register(NEW_PROMOTION_RUN);
        register(DELETE_PROMOTION_RUN);

        register(NEW_VALIDATION_RUN);
        register(NEW_VALIDATION_RUN_STATUS);

        register(PROPERTY_CHANGE);
        register(PROPERTY_DELETE);

    }

    @Override
    public void register(EventType eventType) {
        if (types.containsKey(eventType.getId())) {
            throw new IllegalStateException(format("Event with ID = %s is already registered.", eventType.getId()));
        } else {
            types.put(eventType.getId(), eventType);
        }
    }

    @Override
    public EventType toEventType(String id) {
        EventType eventType = types.get(id);
        if (eventType != null) {
            return eventType;
        } else {
            throw new EventTypeNotFoundException(id);
        }
    }

    @Override
    public Event newProject(Project project) {
        return Event.of(NEW_PROJECT).withProject(project).get();
    }

    @Override
    public Event updateProject(Project project) {
        return Event.of(UPDATE_PROJECT).withProject(project).get();
    }

    @Override
    public Event disableProject(Project project) {
        return Event.of(DISABLE_PROJECT).withProject(project).get();
    }

    @Override
    public Event enableProject(Project project) {
        return Event.of(ENABLE_PROJECT).withProject(project).get();
    }

    @Override
    public Event deleteProject(Project project) {
        return Event.of(DELETE_PROJECT)
                .with("project", project.getName())
                .with("project_id", project.getId().toString())
                .get();
    }

    @Override
    public Event newBranch(Branch branch) {
        return Event.of(NEW_BRANCH).withBranch(branch).get();
    }

    @Override
    public Event updateBranch(Branch branch) {
        return Event.of(UPDATE_BRANCH).withBranch(branch).get();
    }

    @Override
    public Event disableBranch(Branch branch) {
        return Event.of(DISABLE_BRANCH).withBranch(branch).get();
    }

    @Override
    public Event enableBranch(Branch branch) {
        return Event.of(ENABLE_BRANCH).withBranch(branch).get();
    }

    @Override
    public Event deleteBranch(Branch branch) {
        return Event.of(DELETE_BRANCH)
                .withProject(branch.getProject())
                .with("branch", branch.getName())
                .with("branch_id", branch.getId().toString())
                .get();
    }

    @Override
    public Event newBuild(Build build) {
        return Event.of(NEW_BUILD)
                .withBuild(build)
                .get();
    }

    @Override
    public Event updateBuild(Build build) {
        return Event.of(UPDATE_BUILD)
                .withBuild(build)
                .withNoSignature()
                .get();
    }

    @Override
    public Event deleteBuild(Build build) {
        return Event.of(DELETE_BUILD)
                .withBranch(build.getBranch())
                .with("build", build.getName())
                .with("build_id", build.getId().toString())
                .get();
    }

    @Override
    public Event newPromotionLevel(PromotionLevel promotionLevel) {
        return Event.of(NEW_PROMOTION_LEVEL)
                .withPromotionLevel(promotionLevel)
                .get();
    }

    @Override
    public Event imagePromotionLevel(PromotionLevel promotionLevel) {
        return Event.of(IMAGE_PROMOTION_LEVEL)
                .withPromotionLevel(promotionLevel)
                .get();
    }

    @Override
    public Event updatePromotionLevel(PromotionLevel promotionLevel) {
        return Event.of(UPDATE_PROMOTION_LEVEL)
                .withPromotionLevel(promotionLevel)
                .get();
    }

    @Override
    public Event deletePromotionLevel(PromotionLevel promotionLevel) {
        return Event.of(DELETE_PROMOTION_LEVEL)
                .withBranch(promotionLevel.getBranch())
                .with("promotion_level", promotionLevel.getName())
                .with("promotion_level_id", promotionLevel.getId().toString())
                .get();
    }

    @Override
    public Event reorderPromotionLevels(Branch branch) {
        return Event.of(REORDER_PROMOTION_LEVEL)
                .withBranch(branch)
                .get();
    }

    @Override
    public Event newPromotionRun(PromotionRun promotionRun) {
        return Event.of(NEW_PROMOTION_RUN)
                .withPromotionRun(promotionRun)
                .get();
    }

    @Override
    public Event deletePromotionRun(PromotionRun promotionRun) {
        return Event.of(DELETE_PROMOTION_RUN)
                .withBranch(promotionRun.getBuild().getBranch())
                .with(promotionRun.getBuild())
                .with(promotionRun)
                .with(promotionRun.getPromotionLevel())
                .get();
    }

    @Override
    public Event newValidationStamp(ValidationStamp validationStamp) {
        return Event.of(NEW_VALIDATION_STAMP)
                .withValidationStamp(validationStamp)
                .get();
    }

    @Override
    public Event imageValidationStamp(ValidationStamp validationStamp) {
        return Event.of(IMAGE_VALIDATION_STAMP)
                .withValidationStamp(validationStamp)
                .get();
    }

    @Override
    public Event updateValidationStamp(ValidationStamp validationStamp) {
        return Event.of(UPDATE_VALIDATION_STAMP)
                .withValidationStamp(validationStamp)
                .get();
    }

    @Override
    public Event deleteValidationStamp(ValidationStamp validationStamp) {
        return Event.of(DELETE_VALIDATION_STAMP)
                .withBranch(validationStamp.getBranch())
                .with("validation_stamp", validationStamp.getName())
                .with("validation_stamp_id", validationStamp.getId().toString())
                .get();
    }

    @Override
    public Event reorderValidationStamps(Branch branch) {
        return Event.of(REORDER_VALIDATION_STAMP)
                .withBranch(branch)
                .get();
    }

    @Override
    public Event newValidationRun(ValidationRun validationRun) {
        return Event.of(NEW_VALIDATION_RUN)
                .withValidationRun(validationRun)
                .withValidationRunStatus(validationRun.getLastStatus().getStatusID())
                .get();
    }

    @Override
    public Event newValidationRunStatus(ValidationRun validationRun) {
        return Event.of(NEW_VALIDATION_RUN_STATUS)
                .withValidationRun(validationRun)
                .withValidationRunStatus(validationRun.getLastStatus().getStatusID())
                .get();
    }

    @Override
    public <T> Event propertyChange(ProjectEntity entity, PropertyType<T> propertyType) {
        return Event.of(PROPERTY_CHANGE)
                .withRef(entity)
                .with("entity", entity.getProjectEntityType().getDisplayName())
                .with("property", new NameValue(
                        propertyType.getTypeName(),
                        propertyType.getName()
                ))
                .get();
    }

    @Override
    public <T> Event propertyDelete(ProjectEntity entity, PropertyType<T> propertyType) {
        return Event.of(PROPERTY_DELETE)
                .withRef(entity)
                .with("entity", entity.getProjectEntityType().getDisplayName())
                .with("property", new NameValue(
                        propertyType.getTypeName(),
                        propertyType.getName()
                ))
                .get();
    }

    @Override
    public <T extends Configuration<T>> Event newConfiguration(T configuration) {
        return Event.of(NEW_CONFIGURATION)
                .with("configuration", configuration.getName())
                .get();
    }

    @Override
    public <T extends Configuration<T>> Event updateConfiguration(T configuration) {
        return Event.of(UPDATE_CONFIGURATION)
                .with("configuration", configuration.getName())
                .get();
    }

    @Override
    public <T extends Configuration<T>> Event deleteConfiguration(T configuration) {
        return Event.of(DELETE_CONFIGURATION)
                .with("configuration", configuration.getName())
                .get();
    }
}

File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/events/EventFactoryImpl.kt
package net.nemerosa.ontrack.model.events

import net.nemerosa.ontrack.model.events.Event.Companion.of
import net.nemerosa.ontrack.model.events.EventFactory.Companion.DELETE_BRANCH
import net.nemerosa.ontrack.model.events.EventFactory.Companion.DELETE_PROJECT
import net.nemerosa.ontrack.model.events.EventFactory.Companion.DISABLE_BRANCH
import net.nemerosa.ontrack.model.events.EventFactory.Companion.DISABLE_PROJECT
import net.nemerosa.ontrack.model.events.EventFactory.Companion.ENABLE_BRANCH
import net.nemerosa.ontrack.model.events.EventFactory.Companion.ENABLE_PROJECT
import net.nemerosa.ontrack.model.events.EventFactory.Companion.NEW_BRANCH
import net.nemerosa.ontrack.model.events.EventFactory.Companion.NEW_PROJECT
import net.nemerosa.ontrack.model.events.EventFactory.Companion.UPDATE_BRANCH
import net.nemerosa.ontrack.model.events.EventFactory.Companion.UPDATE_PROJECT
import net.nemerosa.ontrack.model.events.EventFactory.Companion.NEW_BUILD
import net.nemerosa.ontrack.model.events.EventFactory.Companion.UPDATE_BUILD
import net.nemerosa.ontrack.model.events.EventFactory.Companion.DELETE_BUILD
import net.nemerosa.ontrack.model.events.EventFactory.Companion.DELETE_CONFIGURATION
import net.nemerosa.ontrack.model.events.EventFactory.Companion.NEW_PROMOTION_LEVEL
import net.nemerosa.ontrack.model.events.EventFactory.Companion.IMAGE_PROMOTION_LEVEL
import net.nemerosa.ontrack.model.events.EventFactory.Companion.UPDATE_PROMOTION_LEVEL
import net.nemerosa.ontrack.model.events.EventFactory.Companion.DELETE_PROMOTION_LEVEL
import net.nemerosa.ontrack.model.events.EventFactory.Companion.REORDER_PROMOTION_LEVEL
import net.nemerosa.ontrack.model.events.EventFactory.Companion.NEW_VALIDATION_STAMP
import net.nemerosa.ontrack.model.events.EventFactory.Companion.IMAGE_VALIDATION_STAMP
import net.nemerosa.ontrack.model.events.EventFactory.Companion.UPDATE_VALIDATION_STAMP
import net.nemerosa.ontrack.model.events.EventFactory.Companion.DELETE_VALIDATION_STAMP
import net.nemerosa.ontrack.model.events.EventFactory.Companion.REORDER_VALIDATION_STAMP
import net.nemerosa.ontrack.model.events.EventFactory.Companion.NEW_PROMOTION_RUN
import net.nemerosa.ontrack.model.events.EventFactory.Companion.DELETE_PROMOTION_RUN
import net.nemerosa.ontrack.model.events.EventFactory.Companion.NEW_CONFIGURATION
import net.nemerosa.ontrack.model.events.EventFactory.Companion.NEW_VALIDATION_RUN
import net.nemerosa.ontrack.model.events.EventFactory.Companion.NEW_VALIDATION_RUN_STATUS
import net.nemerosa.ontrack.model.events.EventFactory.Companion.PROPERTY_CHANGE
import net.nemerosa.ontrack.model.events.EventFactory.Companion.PROPERTY_DELETE
import net.nemerosa.ontrack.model.events.EventFactory.Companion.UPDATE_CONFIGURATION
import net.nemerosa.ontrack.model.events.EventFactory.Companion.UPDATE_VALIDATION_RUN_STATUS_COMMENT
import net.nemerosa.ontrack.model.structure.*
import net.nemerosa.ontrack.model.support.Configuration
import net.nemerosa.ontrack.model.support.NameValue
import org.springframework.stereotype.Service
import java.util.concurrent.ConcurrentHashMap

@Service
class EventFactoryImpl : EventFactory {

    private val types: MutableMap<String, EventType> =
        ConcurrentHashMap()

    init {
        register(NEW_PROJECT)
        register(UPDATE_PROJECT)
        register(DISABLE_PROJECT)
        register(ENABLE_PROJECT)
        register(DELETE_PROJECT)

        register(NEW_BRANCH)
        register(UPDATE_BRANCH)
        register(DISABLE_BRANCH)
        register(ENABLE_BRANCH)
        register(DELETE_BRANCH)

        register(NEW_BUILD)
        register(UPDATE_BUILD)
        register(DELETE_BUILD)

        register(NEW_PROMOTION_LEVEL)
        register(IMAGE_PROMOTION_LEVEL)
        register(UPDATE_PROMOTION_LEVEL)
        register(DELETE_PROMOTION_LEVEL)
        register(REORDER_PROMOTION_LEVEL)

        register(NEW_VALIDATION_STAMP)
        register(IMAGE_VALIDATION_STAMP)
        register(UPDATE_VALIDATION_STAMP)
        register(DELETE_VALIDATION_STAMP)
        register(REORDER_VALIDATION_STAMP)

        register(NEW_PROMOTION_RUN)
        register(DELETE_PROMOTION_RUN)

        register(NEW_VALIDATION_RUN)
        register(NEW_VALIDATION_RUN_STATUS)
        register(UPDATE_VALIDATION_RUN_STATUS_COMMENT)

        register(PROPERTY_CHANGE)
        register(PROPERTY_DELETE)

        register(NEW_CONFIGURATION)
        register(UPDATE_CONFIGURATION)
        register(DELETE_CONFIGURATION)
    }

    final override fun register(eventType: EventType) {
        check(!types.containsKey(eventType.id)) {
            String.format(
                "Event with ID = %s is already registered.",
                eventType.id
            )
        }
        types[eventType.id] = eventType
    }

    override val eventTypes: Collection<EventType>
        get() = types.values

    override fun toEventType(id: String): EventType {
        val eventType = types[id]
        if (eventType != null) {
            return eventType
        } else {
            throw EventTypeNotFoundException(id)
        }
    }

    override fun newProject(project: Project): Event {
        return of(NEW_PROJECT).withProject(project).build()
    }

    override fun updateProject(project: Project): Event {
        return of(UPDATE_PROJECT).withProject(project).build()
    }

    override fun disableProject(project: Project): Event {
        return of(DISABLE_PROJECT).withProject(project).build()
    }

    override fun enableProject(project: Project): Event {
        return of(ENABLE_PROJECT).withProject(project).build()
    }

    override fun deleteProject(project: Project): Event {
        return of(DELETE_PROJECT)
            .with("PROJECT", project.name)
            .with("PROJECT_ID", project.id.toString())
            .build()
    }

    override fun newBranch(branch: Branch): Event {
        return of(NEW_BRANCH).withBranch(branch).build()
    }

    override fun updateBranch(branch: Branch): Event {
        return of(UPDATE_BRANCH).withBranch(branch).build()
    }

    override fun disableBranch(branch: Branch): Event {
        return of(DISABLE_BRANCH).withBranch(branch).build()
    }

    override fun enableBranch(branch: Branch): Event {
        return of(ENABLE_BRANCH).withBranch(branch).build()
    }

    override fun deleteBranch(branch: Branch): Event {
        return of(DELETE_BRANCH)
            .withProject(branch.project)
            .with("BRANCH", branch.name)
            .with("BRANCH_ID", branch.id.toString())
            .build()
    }

    override fun newBuild(build: Build): Event {
        return of(NEW_BUILD)
            .withBuild(build)
            .build()
    }

    override fun updateBuild(build: Build): Event {
        return of(UPDATE_BUILD)
            .withBuild(build)
            .withNoSignature()
            .build()
    }

    override fun deleteBuild(build: Build): Event {
        return of(DELETE_BUILD)
            .withBranch(build.branch)
            .with("BUILD", build.name)
            .with("BUILD_ID", build.id.toString())
            .build()
    }

    override fun newPromotionLevel(promotionLevel: PromotionLevel): Event {
        return of(NEW_PROMOTION_LEVEL)
            .withPromotionLevel(promotionLevel)
            .build()
    }

    override fun imagePromotionLevel(promotionLevel: PromotionLevel): Event {
        return of(IMAGE_PROMOTION_LEVEL)
            .withPromotionLevel(promotionLevel)
            .build()
    }

    override fun updatePromotionLevel(promotionLevel: PromotionLevel): Event {
        return of(UPDATE_PROMOTION_LEVEL)
            .withPromotionLevel(promotionLevel)
            .build()
    }

    override fun deletePromotionLevel(promotionLevel: PromotionLevel): Event {
        return of(DELETE_PROMOTION_LEVEL)
            .withBranch(promotionLevel.branch)
            .with("PROMOTION_LEVEL", promotionLevel.name)
            .with("PROMOTION_LEVEL_ID", promotionLevel.id.toString())
            .build()
    }

    override fun reorderPromotionLevels(branch: Branch): Event {
        return of(REORDER_PROMOTION_LEVEL)
            .withBranch(branch)
            .build()
    }

    override fun newPromotionRun(promotionRun: PromotionRun): Event {
        return of(NEW_PROMOTION_RUN)
            .withPromotionRun(promotionRun)
            .build()
    }

    override fun deletePromotionRun(promotionRun: PromotionRun): Event {
        return of(DELETE_PROMOTION_RUN)
            .withBranch(promotionRun.build.branch)
            .with(promotionRun.build)
            .with("PROMOTION_RUN_ID", promotionRun.id.toString())
            .with(promotionRun.promotionLevel)
            .build()
    }

    override fun newValidationStamp(validationStamp: ValidationStamp): Event {
        return of(NEW_VALIDATION_STAMP)
            .withValidationStamp(validationStamp)
            .build()
    }

    override fun imageValidationStamp(validationStamp: ValidationStamp): Event {
        return of(IMAGE_VALIDATION_STAMP)
            .withValidationStamp(validationStamp)
            .build()
    }

    override fun updateValidationStamp(validationStamp: ValidationStamp): Event {
        return of(UPDATE_VALIDATION_STAMP)
            .withValidationStamp(validationStamp)
            .build()
    }

    override fun deleteValidationStamp(validationStamp: ValidationStamp): Event {
        return of(DELETE_VALIDATION_STAMP)
            .withBranch(validationStamp.branch)
            .with("VALIDATION_STAMP", validationStamp.name)
            .with("VALIDATION_STAMP_ID", validationStamp.id.toString())
            .build()
    }

    override fun reorderValidationStamps(branch: Branch): Event {
        return of(REORDER_VALIDATION_STAMP)
            .withBranch(branch)
            .build()
    }

    override fun newValidationRun(validationRun: ValidationRun): Event {
        return of(NEW_VALIDATION_RUN)
            .withValidationRun(validationRun)
            .withValidationRunStatus(validationRun.lastStatus.statusID)
            .build()
    }

    override fun newValidationRunStatus(validationRun: ValidationRun): Event {
        return of(NEW_VALIDATION_RUN_STATUS)
            .withValidationRun(validationRun)
            .withValidationRunStatus(validationRun.lastStatus.statusID)
            .build()
    }

    override fun updateValidationRunStatusComment(validationRun: ValidationRun): Event {
        return of(UPDATE_VALIDATION_RUN_STATUS_COMMENT)
            .withValidationRun(validationRun)
            .build()
    }

    override fun <T> propertyChange(entity: ProjectEntity, propertyType: PropertyType<T>): Event {
        return of(PROPERTY_CHANGE)
            .withRef(entity)
            .with(
                "PROPERTY",
                NameValue(
                    propertyType.name,
                    propertyType.typeName
                )
            )
            .build()
    }

    override fun <T> propertyDelete(entity: ProjectEntity, propertyType: PropertyType<T>): Event {
        return of(PROPERTY_DELETE)
            .withRef(entity)
            .with(
                "PROPERTY",
                NameValue(
                    propertyType.name,
                    propertyType.typeName
                )
            )
            .build()
    }

    override fun <T : Configuration<T>> newConfiguration(configuration: T): Event {
        return of(NEW_CONFIGURATION)
            .with("CONFIGURATION", configuration.name)
            .build()
    }

    override fun <T : Configuration<T>> updateConfiguration(configuration: T): Event {
        return of(UPDATE_CONFIGURATION)
            .with("CONFIGURATION", configuration.name)
            .build()
    }

    override fun <T : Configuration<T>> deleteConfiguration(configuration: T): Event {
        return of(DELETE_CONFIGURATION)
            .with("CONFIGURATION", configuration.name)
            .with("CONFIGURATION_TYPE", configuration.javaClass.name)
            .build()
    }
}
